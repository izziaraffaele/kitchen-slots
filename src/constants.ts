export enum Ingredients {
  Bacon = 'B',
  Lattuce = 'L',
  Tomato = 'T',
  VeganPatty = 'V',
  Patty = 'P',
}
