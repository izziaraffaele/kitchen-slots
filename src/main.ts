#!/usr/bin/env node
import yargs = require('yargs');

yargs(process.argv.slice(2))
  .usage('Usage: $0 <source>')
  .command(require('./command'))
  .help('h')
  .alias('h', 'help').argv;
