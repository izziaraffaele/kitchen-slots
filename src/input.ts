import fs from 'fs';
import trimEnd from 'lodash/trimEnd';
import chunk from 'lodash/chunk';

import { Ingredients } from './constants';

const inventoryKeyMap = [
  Ingredients.Patty,
  Ingredients.Lattuce,
  Ingredients.Tomato,
  Ingredients.VeganPatty,
  Ingredients.Bacon,
];

const parseString = (input: string) => {
  const [firstRow, ...otherRows] = trimEnd(input, '\n')
    .split('\n') // split string to lines
    .map((e) => e.trim()) // remove white spaces for each line
    .map((e) => e.split(',').map((e) => e.trim()));

  if (!firstRow) {
    throw new Error('INVALID_INPUT');
  }

  const merchant = {
    id: firstRow[0],
    workstations: chunk(firstRow.slice(1, 7), 2).map(([capacity, time]) => ({
      id: capacity[1],
      capacity: parseInt(capacity[0], 10),
      time: parseInt(time, 10),
    })),
    inventory: firstRow.slice(7).reduce((carry, item, index) => {
      const itemId = inventoryKeyMap[index];
      carry[itemId] = parseInt(item, 10);
      return carry;
    }, {} as Record<string, number>),
  };

  const orders = otherRows.map(([merchantId, date, id, ...items]) => ({
    id,
    merchantId,
    date: new Date(date),
    items: items,
  }));

  return { merchant, orders };
};

const parseSource = (source: string) => {
  if (fs.existsSync(source)) {
    return parseString(fs.readFileSync(source).toString());
  }
};

export default { parseSource, parseString };
