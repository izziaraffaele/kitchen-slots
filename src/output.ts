import { IIngredients, IOrder } from './core/interfaces';
import { Ingredients } from './constants';

export interface Response {
  merchant: string;
  total: number;
  inventory: IIngredients;
  orders: IOrder[];
}

export const send = (data: Response) => {
  const rows: string[] = [];

  data.orders.forEach((order) => {
    rows.push(
      [order.merchantId, order.id, order.status, order.processingTime].join(',')
    );
  });

  rows.push([data.merchant, 'TOTAL', data.total].join(','));

  rows.push(
    [
      data.merchant,
      'INVENTORY',
      data.inventory[Ingredients.Patty],
      data.inventory[Ingredients.Lattuce],
      data.inventory[Ingredients.Tomato],
      data.inventory[Ingredients.VeganPatty],
      data.inventory[Ingredients.Bacon],
    ].join(',')
  );

  rows.push('\n');

  console.log(rows.join('\n'));
};
