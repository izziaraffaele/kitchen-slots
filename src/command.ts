import yargs = require('yargs');

import Input from './input';
import * as Merchant from './core/merchant';
import * as Output from './output';

interface Arguments extends yargs.Arguments {
  source: string;
}

export const command = ['run <source>', '$0'];

export const desc = 'Run the process';

export const builder = (yargs: any) => {
  yargs.positional('source', {
    type: 'string',
    describe: 'the relative path to the file containing input data',
  });
};

export const handler = function (argv: Arguments) {
  const input = Input.parseSource(argv.source);

  if (!input?.merchant) {
    throw new Error('Invalid input');
  }

  const merchant = Merchant.create(
    input.merchant.inventory,
    input.merchant.workstations
  );

  const processedOrders = merchant.process(
    input.orders.map((order) => ({
      ...order,
      status: 'PENDING',
      processingTime: 0,
    }))
  );

  Output.send({
    merchant: input.merchant.id,
    orders: processedOrders,
    inventory: merchant.getInventory(),
    total: merchant.getTotalProcessingTime(),
  });
};
