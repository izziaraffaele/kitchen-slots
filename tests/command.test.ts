import yargs = require('yargs');

import * as command from '../src/command';

const spy = jest.spyOn(global.console, 'log');

describe('CLI', () => {
  beforeEach(() => {
    spy.mockClear();
  });

  it('returns help output', async () => {
    // Initialize parser using the command module
    const parser = yargs.command(command as any).help();

    // Run the command module with --help as argument
    const output: string = await new Promise((resolve) => {
      parser.parse('--help', (err: any, argv: yargs.Arguments, output: any) => {
        resolve(output);
      });
    });

    // Verify the output is correct
    expect(output).toEqual(expect.stringContaining(command.desc));
  });

  it('should return the expected output', async () => {
    // Initialize parser using the command module
    const parser = yargs.command(command as any).help();

    // Run the command module with --help as argument
    await new Promise((resolve) => {
      parser.parse(
        'tests/fixtures/test-input.csv',
        (err: any, argv: yargs.Arguments, output: any) => {
          resolve(output);
        }
      );
    });

    expect(spy.mock.calls[0][0]).toMatchSnapshot();
  });
});
