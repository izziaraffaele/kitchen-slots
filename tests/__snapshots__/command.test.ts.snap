// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI should return the expected output 1`] = `
"R1,O1,ACCEPTED,5
R1,O2,ACCEPTED,9
R1,O6,ACCEPTED,9
R1,O3,ACCEPTED,9
R1,O7,ACCEPTED,9
R1,O4,ACCEPTED,12
R1,O8,ACCEPTED,9
R1,O9,ACCEPTED,9
R1,O5,ACCEPTED,5
R1,O10,ACCEPTED,4
R1,O11,ACCEPTED,4
R1,O12,ACCEPTED,4
R1,TOTAL,29
R1,INVENTORY,73,161,151,78,80

"
`;
